xgboost_ray/main.py

```diff
@@ -826,7 +884,18 @@ def train(params: Dict,
     checkpoint = _Checkpoint()  # Keep track of latest checkpoint
     current_results = {}  # Keep track of additional results
     actors = [None] * ray_params.num_actors  # All active actors
+
+    # Create the Queue and Event actors.
     queue, stop_event = _create_communication_processes()
+
+    if not ray_params.elastic_training:
+        strategy = "PACK" if added_tune_callback else "SPREAD"
+        pg = _create_placement_group(cpus_per_actor, gpus_per_actor,
+                                     ray_params.resources_per_actor,
+                                     ray_params.num_actors, strategy)
```

Comment:
The main question I have here is if we always want to create placement groups. I'm not familiar enough with them to know the benefits and drawbacks. I get that we would want them with Tune, but without Tune and without elastic training, what is the benefit of using placement groups? Does SPREAD give us any benefits? Do placement groups have any significant overhead?

Reply:
@richardliaw @rkooo567 please feel free to add more info/correct me if I'm wrong here:

One of the benefits of placement groups is **atomic resource reservation** which is useful on a non-autoscaling cluster. With placement groups, if all the required resources cannot be reserved at the same time, then execution will fail (unless autoscaling is enabled). Without placement groups, some of the training workers will be scheduled, and others will not, causing execution to hang I believe. Note that if we are on an autoscaling cluster, both cases should trigger autoscaling, so there should not be a difference between the two.

As for the actual strategy, PACK vs. SPREAD vs. no placement groups should not make a difference if the entire Ray cluster is fully utilized. But if that's not the case, then SPREAD would improve fault tolerance since it minimizes the number of workers that fail when a node goes down (https://docs.ray.io/en/master/auto_examples/placement-group.html#improve-fault-tolerance). However, this comes at a cost of increased inter-node communication. So SPREAD prioritizes fault tolerance over reducing network communication. PACK does the complete opposite and prioritizes reducing network communication over fault tolerance. Not using placement groups does not make any prioritization, and the placement behavior is not deterministic.

As for additional overheads, I don't know the particulars here, but I'm sure at the very least there is some accounting overhead. But I believe that this overhead only occurs for scheduling decisions when actors are created. Once all of the training workers have been placed, actual training times should not increase.

Reply:
Overall, I think for the non-tune case SPREAD is what we want here since it will improve fault tolerance. But, I'm also OK if you think we should discuss this further, and we can disable placement groups for the non-tune case. It's only a 1 line change to toggle this.

Reply:
We can also add an environment variable to toggle this, which will probably make it easier to run benchmarking with/without placement groups.

Reply:
I added an environment variable to toggle this for your benchmarking tomorrow.

Reply:
Awesome. Thanks for the thorough explanation. It sounds like placement groups are generally desired default behavior. I'll try it out tomorrow. If it works I don't have anything else to add and we should be ready to merge.